builder/builder.go

```diff
@@ -26,26 +28,59 @@ type Builder struct {
 	ownAddress felt.Felt
 	privKey    *ecdsa.PrivateKey
 
-	bc       *blockchain.Blockchain
-	vm       vm.VM
-	newHeads *feed.Feed[*core.Header]
-	log      utils.Logger
+	bc        *blockchain.Blockchain
+	vm        vm.VM
+	newHeads  *feed.Feed[*core.Header]
+	log       utils.Logger
+	blockTime time.Duration
+
+	listener EventListener
 }
 
-func New(privKey *ecdsa.PrivateKey, ownAddr *felt.Felt, bc *blockchain.Blockchain, builderVM vm.VM, log utils.Logger) *Builder {
+func New(privKey *ecdsa.PrivateKey, ownAddr *felt.Felt, bc *blockchain.Blockchain, builderVM vm.VM,
+	blockTime time.Duration, log utils.Logger,
+) *Builder {
 	return &Builder{
 		ownAddress: *ownAddr,
 		privKey:    privKey,
+		blockTime:  blockTime,
+		log:        log,
+		listener:   &SelectiveListener{},
 
 		bc:       bc,
 		vm:       builderVM,
 		newHeads: feed.New[*core.Header](),
 	}
 }
 
+func (b *Builder) WithEventListener(l EventListener) *Builder {
+	b.listener = l
+	return b
+}
+
+// Run takes ownership of the Builder.
 func (b *Builder) Run(ctx context.Context) error {
-	<-ctx.Done()
-	return nil
+	ticker := time.NewTicker(b.blockTime)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			pending, err := b.bc.Pending()
+			if err != nil {
+				return fmt.Errorf("get pending: %v", err)
+			}
+			pending.Block.SequencerAddress = &b.ownAddress
+			if err = b.bc.Finalise(&pending, b.Sign); err != nil {
+				return fmt.Errorf("finalise: %v", err)
+			}
+			b.log.Infow("Finalised block", "number", pending.Block.Number, "hash", pending.Block.Hash.ShortString())
+			if b.listener.OnBlockFinalised(pending.Block.Header) {
+				return nil
+			}
```

Comment:
These are just callbacks, not hooks. So they shouldn't affect execution. To terminate the builder execution, we should cancel `ctx`.

Reply:
That makes it pretty difficult to test. If we allow them to be hooks, we can execute a specific number of blocks.

Reply:
I tried other packages with a `Clock` interface that we can mock, but they all brought complexity we don't need imo.